### PR TITLE
ddl:fix create partitioned table with bigint column fail

### PR DESCRIPTION
**Pull request by ciscoxll**
### What problem does this PR solve? 

Create partitioned table when partitioning key is of type `bigint` unsigned will fail.

* Fix issue #7498

### What is changed and how it works?

*  The partitioning key must be an integer, or an expression that returns an integer.
* For `PartitionRangeValue` range comparison.
   * `Unsigned bigint` was converted to `uint64` handle.
   * `Bigint` was converted to `int64` handle.

mysql
```sql
mysql> create table t1 (a bigint unsigned not null) partition by range(a) (
    ->   partition p0 values less than (10),
    ->   partition p1 values less than (100),
    ->   partition p2 values less than (1000),
    ->   partition p3 values less than (18446744073709551000),
    ->   partition p4 values less than (18446744073709551614)
    -> );
Query OK, 0 rows affected (0.07 sec)

mysql> insert into t1 values (5),(15),(105),(1005);
Query OK, 4 rows affected (0.00 sec)
Records: 4  Duplicates: 0  Warnings: 0

mysql> insert into t1 values (18446744073709551000+1);
Query OK, 1 row affected (0.00 sec)

mysql> insert into t1 values (18446744073709551614-1);
Query OK, 1 row affected (0.00 sec)
```

master branch
```sql
mysql> create table t1 (a bigint unsigned not null) partition by range(a) (
    ->   partition p0 values less than (10),
    ->   partition p1 values less than (100),
    ->   partition p2 values less than (1000),
    ->   partition p3 values less than (18446744073709551000),
    ->   partition p4 values less than (18446744073709551614)
    -> );
ERROR 1493 (HY000): VALUES LESS THAN value must be strictly increasing for each partition
```
### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by ciscoxll** [comment]
@tiancaiamao @winkyao @zimulala PTAL .

File: ddl/partition.go [hunk 1]
```diff
@@ -187,13 +187,22 @@ func checkCreatePartitionValue(ctx sessionctx.Context, tblInfo *model.TableInfo,
 	if strings.EqualFold(defs[len(defs)-1].LessThan[0], partitionMaxValue) {
 		defs = defs[:len(defs)-1]
 	}
-	var prevRangeValue int64
+
+	var offset int
+	for i, col := range cols {
```
Comment (zimulala) [inline ddl/partition.go]: Cloud we use "FindCol" directly?

File: ddl/partition.go [hunk 2]
```diff
@@ -206,9 +208,14 @@ func checkCreatePartitionValue(ctx sessionctx.Context, tblInfo *model.TableInfo,
 			prevRangeValue = currentRangeValue
 			continue
 		}
-
-		if currentRangeValue <= prevRangeValue {
-			return errors.Trace(ErrRangeNotIncreasing)
+		if col.Tp == mysql.TypeLonglong && mysql.HasUnsignedFlag(col.Flag) {
```
Comment (tiancaiamao) [inline ddl/partition.go]: `col` maybe `nil` here ?

File: ddl/partition.go [hunk 3]
```diff
@@ -187,13 +187,15 @@ func checkCreatePartitionValue(ctx sessionctx.Context, tblInfo *model.TableInfo,
 	if strings.EqualFold(defs[len(defs)-1].LessThan[0], partitionMaxValue) {
 		defs = defs[:len(defs)-1]
 	}
-	var prevRangeValue int64
+
+	col := table.FindCol(cols, strings.ToUpper(pi.Expr))
```
Comment (tiancaiamao) [inline ddl/partition.go]: FindCol use a case-insensitive search, so `strings.ToUpper` is not necessary

**Comment by ciscoxll** [comment]
@tiancaiamao @winkyao @zimulala PTAL .

File: ddl/partition.go [hunk 4]
```diff
@@ -388,3 +406,15 @@ func checkConstraintIncludePartKey(partkeys []string, constraints map[string]str
 	}
 	return true
 }
+
+// FindRangePartitionColTp find the type of the partitioning key column type.
```
Comment (zhexuany) [inline ddl/partition.go]: s/FindRangePartitionColTp find the type of the partitioning key column type./FindRangePartitionColTp finds the type of the partitioning key column type.

File: ddl/partition.go [hunk 5]
```diff
@@ -217,23 +224,34 @@ func checkCreatePartitionValue(ctx sessionctx.Context, tblInfo *model.TableInfo,
 
 // getRangeValue gets an integer from the range value string.
 // The returned boolean value indicates whether the input string is a constant expression.
-func getRangeValue(ctx sessionctx.Context, tblInfo *model.TableInfo, str string) (int64, bool, error) {
-
-	if value, err := strconv.ParseInt(str, 10, 64); err == nil {
-		return value, false, nil
-	}
+func getRangeValue(ctx sessionctx.Context, tblInfo *model.TableInfo, str string, ifUnsignedBigint bool) (interface{}, bool, error) {
```
Comment (tiancaiamao) [inline ddl/partition.go]: s/ifUnsignedBigint/unsignedBigint

File: ddl/partition.go [hunk 6]
```diff
@@ -388,3 +406,15 @@ func checkConstraintIncludePartKey(partkeys []string, constraints map[string]str
 	}
 	return true
 }
+
+// FindRangePartitionColTp finds the type of the partitioning key column type.
+// The returned boolean value indicates whether the partitioning key column type is unsigned bigint type.
+func FindRangePartitionColTp(cols []*table.Column, pi *model.PartitionInfo) bool {
```
Comment (tiancaiamao) [inline ddl/partition.go]: s/FindRangePartitionColTp/isRangePartitionColUnsignedBigint

Comment (tiancaiamao) [reply ddl/partition.go]: Find column type returns a boolean is weird.

**Comment by ciscoxll** [comment]
@tiancaiamao @winkyao @zimulala PTAL .

**Comment by tiancaiamao** [comment]
LGTM

**Comment by ciscoxll** [comment]
@crazycs520 @winkyao @zimulala PTAL .

File: ddl/partition.go [hunk 7]
```diff
@@ -217,23 +224,34 @@ func checkCreatePartitionValue(ctx sessionctx.Context, tblInfo *model.TableInfo,
 
 // getRangeValue gets an integer from the range value string.
 // The returned boolean value indicates whether the input string is a constant expression.
-func getRangeValue(ctx sessionctx.Context, tblInfo *model.TableInfo, str string) (int64, bool, error) {
-
-	if value, err := strconv.ParseInt(str, 10, 64); err == nil {
-		return value, false, nil
-	}
+func getRangeValue(ctx sessionctx.Context, tblInfo *model.TableInfo, str string, unsignedBigint bool) (interface{}, bool, error) {
```
Comment (crazycs520) [inline ddl/partition.go]: how about return uint64 not interface{}?

Comment (ciscoxll) [reply ddl/partition.go]: @crazycs520 Because it might be `int64` or `uint64`.

Comment (crazycs520) [reply ddl/partition.go]: int64 can use  `uint64( int64Variable_name )` conver to uint64.

Comment (ciscoxll) [reply ddl/partition.go]: @crazycs520 Conversion from signed `int64` to `uint64` can be an error.

**Comment by ciscoxll** [comment]
@crazycs520 @winkyao @zimulala PTAL .

**Pull request merged**